Assure minimum bounding box size before starting AI inference

knollengewaechs opened the pull request:
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- abc

### TODOs:
- [ ] ...

### Issues:
- fixes #

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

knollengewaechs commented on frontend/javascripts/oxalis/view/action-bar/starting_job_modals.tsx:
```diff
@@ -694,6 +708,53 @@ function useCurrentlySelectedBoundingBox(
+    return true;
+  }
+  const datasetExtent: Vector3 = [
+    colorLayer.boundingBox.width,
+    colorLayer.boundingBox.height,
+    colorLayer.boundingBox.depth,
```
check right order

knollengewaechs replied:
:heavy_check_mark:

knollengewaechs commented on conf/application.conf:
```diff
@@ -151,8 +151,8 @@ features {
   # to enable jobs for local development, use "yarn enable-jobs" to also activate it in the database
-  jobsEnabled = false
-  voxelyticsEnabled = false
+  jobsEnabled = true
+  voxelyticsEnabled = true
```
:rotating_light:

daniel-wer commented on frontend/javascripts/oxalis/view/action-bar/starting_job_modals.tsx:
```diff
@@ -75,6 +75,26 @@ import { isBoundingBoxExportable } from "./download_modal_view";
+};
+
+const MIN_BBOX_EXTENT_FOR_MITO_SEGMENTATION: MinimalBoundingBoxExtent = {
+  minExtent: [16, 16, 4],
```
@AndreaNathansen Do you know what the minimum bounding box and or dataset size for the mito model in wk is? See https://github.com/scalableminds/webknossos/issues/8198 for context

AndreaNathansen replied:
I don't know exactly, would have to try it out.

Since it predicts slice-wise, maybe one slice in z direction would be sufficient, but this could fail during the median pooling afterwards.

For x and y direction, the MitoNet paper says "the output from the encoder was 16x smaller than the input", so maybe 16 vx is the limit here, but that would have to be tried as well.

knollengewaechs replied:
is there a way to find out the voxel sizes of the finest available mag of the training dataset? i think that would need to be hardcoded as well, right? or is this a constant across all datasets? I found the scale 11.24 × 11.24 × 28 nm³/voxel in multiple datasets but that could be a coincidence, didnt find anything explicit in the docs

daniel-wer replied:
Sorry, my statement was ambiguous. I think you don't need to know the voxel size of the finest available mag of the training dataset. Contrary to the neuron inference (where the mag of the current dataset that most closely resembles the mag the model was trained in, is chosen for inference), for the mito inference, the finest available mag _of the current dataset_ is chosen for inference. So there is no need to do any mag selection according to the mean training voxel size.

knollengewaechs replied:
oh I see, thank you :D

knollengewaechs replied:
@AndreaNathansen can you find this out or is there anything I can do?

knollengewaechs replied:
thank you for finding this out! actually both is relevant IMO, I am asserting that both the bounding box of the segmentation and the overall dataset are big enough. for nuclei and neuron segmentation I am asserting that the DS is double the size of the minimal bounding box

daniel-wer commented on frontend/javascripts/oxalis/view/action-bar/starting_job_modals.tsx:
```diff
@@ -694,6 +714,54 @@ function useCurrentlySelectedBoundingBox(
+    if (bboxExtent[i] < minExtentInMag1[i]) {
+      const boundingBoxOrDSMessage = bboxOrDS === "bbox" ? "bounding box" : "dataset";
+      Toast.error(
+        `The ${boundingBoxOrDSMessage} is too small. Please select a ${boundingBoxOrDSMessage} with the minimal extent ${minExtentInMag1} Vx.`,
```
```suggestion
        `The ${boundingBoxOrDSMessage} is too small. Please select a ${boundingBoxOrDSMessage} with the minimal extent ${minExtentInMag1} vx.`,
```

daniel-wer commented on frontend/javascripts/oxalis/view/action-bar/starting_job_modals.tsx:
```diff
@@ -694,6 +714,54 @@ function useCurrentlySelectedBoundingBox(
+  if (isBBoxTooSmall(datasetExtent, segmentationType, "dataset")) {
+    return true;
+  }
+  if (isBBoxTooSmall(bbox, segmentationType)) {
```
Might be personal preference, but if you build the `datasetExtent` from the `colorLayer.boundingBox` here anyways, you could do the same for `bbox` and get rid of the special casing in the `isBBoxTooSmall` function (`const bboxExtent = bbox.length === 6 ? bbox.slice(3) : bbox;`).

daniel-wer reviewed (commented):
Mostly LGTM, but two details are missing:
1. Mirroring the automatic mag selection of the worker and incorporating that in the computation
2. Checking the sizes for the nuclei inferral

Please see my comments :)

knollengewaechs commented on frontend/javascripts/oxalis/view/action-bar/starting_job_modals.tsx:
```diff
@@ -75,6 +75,26 @@ import { isBoundingBoxExportable } from "./download_modal_view";
+// See https://github.com/scalableminds/webknossos/issues/8198#issuecomment-2782684436
+const MIN_BBOX_EXTENT_FOR_NEURON_SEGMENTATION: MinimalBoundingBoxExtent = {
+  minExtent: [16, 16, 4],
+  mag: [1, 1, 1],
```
I am confused. What do you mean with `The extent values are in the mag that is predicted in.`? I thought that means that the values have to be scaled to mag1, hence I included the mag for future cases where its not mag1. 
And am I understanding you correctly that the frontend should mirror the computation of the mag selection based on the mean voxel size (that needs to be hardcoded then), rather than hardcoding the bbox the model was trained on?

daniel-wer replied:
Mmh almost I think. It's correct that the mean voxels size the model was trained on needs to be hardcoded, but the minimum extent needs to be available in the mag the model predicts in _for the current dataset_. To find out which mag that is (i.e. which mag of the current dataset most closely resembles the mag the model was trained in) the mag selection code from the worker needs to be mirrored. Once the mag is computed, the min_extent needs to be multiplied by the mag to get the minimum extent in mag 1. That can the be used to validate the bbox and dataset extents.
Does that make sense? Otherwise, let's have a quick call to discuss :)

knollengewaechs commented:
@daniel-wer I hope to have addressed your feedback and implemented all missing parts. I mirrored the voxelytics code and added missing constants. I am excited to hear what you think! Due to the newly added algorithm, I am afraid there were quite some changes after your last review.

daniel-wer commented on frontend/javascripts/viewer/view/action-bar/starting_job_modals.tsx:
```diff
@@ -712,28 +726,64 @@ function useCurrentlySelectedBoundingBox(
+    Number.POSITIVE_INFINITY,
+  ];
+
+  const datasetScaleInNm = convertVoxelSizeToUnit(datasetScaleMag1);
```
I agree that default arguments can be very helpful, but in this case it makes it harder to understand the code, I'd say. The function name doesn't indicate which unit the voxel size is converted to so one needs to look at the function definition to make sense of it. I'd suggest to remove the "nm" parameter default and explicitly include it in the function calls (same for the tests).

daniel-wer commented on frontend/javascripts/viewer/view/action-bar/starting_job_modals.tsx:
```diff
@@ -712,28 +726,64 @@ function useCurrentlySelectedBoundingBox(
+    const diff = datasetScaleInNm.map(
+      (dim, i) => Math.log2(dim * mag[i]) - Math.log2(modelScale[i]),
+    );
+    if (Math.abs(bestDifference[0]) > Math.abs(diff[0])) {
```
I wanted to comment that all dimensions of the difference should be taken into account, but then I saw that the original code does the same :see_no_evil: Therefore, this should be left as-is to mirror the voxelytics code. Once that's improved we'll need to change it here as well.

daniel-wer commented on frontend/javascripts/viewer/view/action-bar/starting_job_modals.tsx:
```diff
@@ -712,28 +726,64 @@ function useCurrentlySelectedBoundingBox(
+  }
+  const maxDistance = Math.max(...bestDifference);
+  const resultText = `Using mag [${closestMagOfCurrentDS}]. This results in an effective scale of [${datasetScaleInNm.map((scale, i) => Math.round(scale * closestMagOfCurrentDS[i]))}] (compared to scale [${modelScale.map((scale) => Math.round(scale))}] used during training).`;
+  if (maxDistance > Math.log2(2)) {
```
```suggestion
  if (maxDistance > Math.log(2)) {
```

daniel-wer commented on frontend/javascripts/viewer/view/action-bar/starting_job_modals.tsx:
```diff
@@ -712,28 +726,64 @@ function useCurrentlySelectedBoundingBox(
+
+  for (const mag of colorLayer.resolutions) {
+    const diff = datasetScaleInNm.map(
+      (dim, i) => Math.log2(dim * mag[i]) - Math.log2(modelScale[i]),
```
Pythons and numpys log functions use the natural logarithm.

```suggestion
      (dim, i) => Math.log(dim * mag[i]) - Math.log(modelScale[i]),
```

daniel-wer commented on frontend/javascripts/viewer/view/action-bar/starting_job_modals.tsx:
```diff
@@ -712,28 +726,64 @@ function useCurrentlySelectedBoundingBox(
+      closestMagOfCurrentDS = mag;
+    }
+  }
+  const maxDistance = Math.max(...bestDifference);
```
The original line is `np.max(np.abs(closest_diff))`. I think the `abs` is important because of negative differences and should be added here as well.

knollengewaechs replied:
I changed it so that the original calculation of the diff simply takes the absolute (l 750
`const diff = datasetScaleInNm.map((dim, i) => Math.abs(Math.log(dim * mag[i]) - Math.log(modelScale[i])));`. Then the diff is always positive and the absolute can only be taken once. This means that this code differs a bit from the vx code, but it does anyway because of the different ways numpy and TS handle arrays. If you think they should be as similar as possible, please let me know, then I will change it back.

daniel-wer replied:
Great solution :+1:

daniel-wer reviewed (commented):
Very nice, mostly LGTM, but please see my small suggestions. I'll let you know if anything comes up during testing.

daniel-wer reviewed (commented):
Testing went nicely, just added two more comments :)
